contrib/insights/src/insights.cxx

```diff
@@ -0,0 +1,215 @@
+/*
+ * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"). You
+ * may not use this file except in compliance with the License. A copy of
+ * the License is located at
+ *
+ *      http://aws.amazon.com/apache2.0/
+ *
+ * or in the "license" file accompanying this file. This file is
+ * distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+ * ANY KIND, either express or implied. See the License for the specific
+ * language governing permissions and limitations under the License.
+ */
+
+#include "module.h"
+#include "mlio.h"
+#include "mlio/parser.h"
+#include "mlio/util/number.h"
+#include "mlio/util/string.h"
+#include "tbb/tbb.h"
+
+#include <pybind11/stl_bind.h>
+#include <pybind11/stl.h>
+#include <pybind11/functional.h>
+
+#include "insights.h"
+#include "column_statistics.h"
+#include "column_analyzer.h"
+
+#include <exception>
+#include <functional>
+#include <string>
+#include <cmath>
+#include <limits>
+#include <map>
+#include <utility>
+
+namespace py = pybind11;
+
+using namespace tbb;
+using namespace pybind11::literals;
+
+namespace insights {
+namespace detail {
+
+
+pybind11::object
+analyze_dataset(
+    mlio::data_reader * reader_,
+    std::unordered_set<std::string> const *null_like_values,
+    std::unordered_set<size_t> const *capture_columns,
+    size_t max_capture_count)
+{   
+    // Stores the data that our parallel function will operate on.
+    mlio::intrusive_ptr<mlio::example> exm = nullptr;
+    std::vector<column_analysis> columns;
+
+    // Analyzer that is used in the parallel function.
+    auto analyzer = column_analyzer(&exm, 
+                                    &columns, 
+                                    null_like_values,
+                                    capture_columns,
+                                    max_capture_count);
+
+    // First set up the column_analsis objects to store the
+    // collected information.
+    exm = reader_->peek_example();
+    if (exm == nullptr) {
+        throw pybind11::stop_iteration("Reached end of file without reading an example");
+    }
+    if (columns.empty()) {
+        columns.reserve(exm->features().size());
+        for (mlio::feature_desc desc : exm->get_schema().descriptors()) {
```

Comment:
You should access the elements by reference; otherwise this causes an unnecessary copy operation.
```
for (mlio::feature_desc const &desc : exm->get_schema().descriptors()) {
  ...
}
```